### ISSUE TITLE
[Feature Request] - Show Uncommitted Changes in Commit History

### ISSUE THREAD

**Issue by cwhenderson20** [ForkIssues/Tracker#308]
I'm loving Fork so far! One feature that I do miss (coming from SourceTree) is the ability to see current uncommitted changes within the history view:

<img width="594" alt="screen shot 2018-06-11 at 2 42 08 pm" src="https://user-images.githubusercontent.com/5496110/41250755-d72701aa-6d85-11e8-8474-a5c4c7a795f2.png">

For me, it makes more sense than separating out changes into their own section because ultimately, those changes will become part of the history once they're committed. It also reduces the amount of context switching as I can see past commit messages when I'm ready to commit changes.

I'm sure not everyone agrees though, so some kind of preference such as "Show uncommitted changes with history" would be a great compromise.

Thanks for a great product!

**Comment by bwl21** [comment]
I Repeat ja Comment to #107 in order to second this ...

Very cool. Now we can see that we are on a detached head.
I am afraid I propose to do more:

Here is a situation in fork,

![screenshot_1046](https://user-images.githubusercontent.com/2022976/33492675-0d5ac626-d6be-11e7-8a8b-7ea718bc0efc.jpg)

* love the idea to visualize the stashes in the history tree. 
* I would also like to see the HEAD visualized in the history tree
* I would also like the "uncommitted changes" in the history tree

Here is what I expected (Sorry Atlassian, I shameless copy Sourcetree's beahavior). But this visualization is the coolest approach I have seen for this.

![screenshot_1045](https://user-images.githubusercontent.com/2022976/33492669-06e4857a-d6be-11e7-91bb-84e0647fefb1.jpg)

**Comment by DanPristupov** [comment]
What should be shown in details view when the 'uncommitted changes' item is selected?

**Comment by bwl21** [comment]
It should be the same content shown by Fork when I click on "changes" in the sidebar.

**Comment by DanPristupov** [comment]
I think it will be very confusing to show the changes in two views, but allow to commit only in one.

**Comment by bwl21** [comment]
If you show the contents of "changes" in the details view and allow to state, unstage, commit as you do in the "changes" view. It is not that confusing.

you could even switch to "changes" if user clicks on "uncommited changes". The primary reason for the request is that I can see in the timline if anything is commited properly. As the number of uncommited changes-Tag on "changes" is not reliable, I always have to click on "changes" to check the status of all files.

**Comment by omarn3tworkcom** [comment]
I'm glad this is currently a thing, I love the idea of "Uncommitted changes" or "Working copy changes" as the top item in the graph. Several git clients do it and it just feels right. 

This way, the Changes section can be included in that top item when selected in the graph and the Changes section can be removed all together.

So, it's just a matter of moving the section to another place, not duplicating it.

**Comment by omarn3tworkcom** [comment]
Here is an example of how the two view would look:
[History.png](https://cdn.pbrd.co/images/HuQ2Qt9.png)
[Commit.png](https://cdn.pbrd.co/images/HuQ3x78.png)

**Comment by maksimovic** [comment]
Maybe it's just a relic from SourceTree's habits, and a bit too much to expect the flow to be replicated 1:1 in Fork. I've used SourceTree for years, then switched to Fork and I must admit that I did miss its "uncommited changes" a bit. However, as soon as I figured out that `CMD`+`1` shows exactly that, it became natural to toggle between the history view (`CMD`+`2`) and uncommitted changes by keyboard.

**Comment by omarn3tworkcom** [comment]
There is no problem in usability however, it's a matter of less keys/clicks as much as possible without losing track of all the information.

**Comment by dalewking** [comment]
Having another screen to see changes is not equivalent functionality. In the history view I can select 2 commits and see the difference between those commits. In SourceTree I can also select the uncommitted changes and another commit and see the difference between my current workspace and some arbitrary commit. No way to do that in Fork. That is a feature I have used often in SourceTree. I can tell you that feature is somewhat buggy in SourceTree in that if you go edit the files and then go back to sourcetree it will not update the diff automatically and you have to go select them again to see the updated diff, but I live with that minor annoyance as this feature is so useful.

The question of what to show when selecting it is simple it just shows the difference between your workspace and the last commit, just like what git diff would show.

**Comment by dalewking** [comment]
Actually what ST shows at the bottom is mostly the same view that is under File status (which is the same as your Changes page) just without the controls for doing a commit. So it has Unstaged changes list, Staged changes list, and diff, but not the commit message, etc.

**Comment by ibushong** [comment]
+1

I just switched over to Fork and I wish it had this feature too (I used this all the time in ST). Maybe I just need to get used to a different workflow, but I still see an argument for adding this in.

**Comment by bwl21** [comment]
To be honest, I now got used to it and no longer request it urgently. It seems that newbies want it - and I am no longer a newbie :-)

**Comment by maksimovic** [comment]
Come on, <kbd>Cmd</kbd>+<kbd>1</kbd> is way faster than hunting the top row in the commit history 😼

**Comment by dalewking** [comment]
> To be honest, I now got used to it and no longer request it urgently. It seems that newbies want it - and I am no longer a newbie :-)

To be honest, I could never get used to it and no longer request it urgently.  It seems that users want it - and I am no longer a user

**Comment by omarn3tworkcom** [comment]
> I'm no longer a user 
Made me laugh

> Come on, Cmd+1 is way faster than hunting the top row in the commit history 😼
No need for a shortcut is even faster you know

**Comment by brechtm** [comment]
It would be useful to have the uncommited changes show up in the commit history* to be able to diff between e.g. an arbitrary commit/stash entry and the uncommitted changes. Or his this already possible in another way?

_*_ perhaps even showing staged/non-staged changes as separate entries

**Comment by bwl21** [comment]
u can stash the current changes. then they show up in the commit history and you can compare them

**Comment by brechtm** [comment]
> u can stash the current changes. then they show up in the commit history and you can compare them

True, but that requires an extra step and makes you loose staged changes (or rather the distinction between staged/unstaged changes).

**Comment by bwl21** [comment]
> loose staged changes (or rather the distinction between staged/unstaged changes).

agree, for this reason, I have  the same request. My proposal is a workaround - better than nothing

**Comment by r-owen** [comment]
> I think it will be very confusing to show the changes in two views, but allow to commit only in one.

I think you should try SourceTree before dismissing this suggestion. I have never found it confusing, and I find it very useful to be able to compare the current state of a file (local changes) to older commits. The lack of this feature in Fork is the main reason I still use SourceTree instead of Fork (the other being the slightly-to-small font).

**Comment by DanPristupov** [comment]
@r-owen

> I think you should try SourceTree before dismissing this suggestion.

About 7 years ago I was looking for a git client for Mac. I tried ST, couldn't figure out how to work with it and in desperation I decided to make Fork.

> I find it very useful to be able to compare the current state of a file (local changes) to older commits. 

Right click on a commit -> `Compare to Local Changes`

<img width="356" alt="Screenshot 2023-05-09 at 21 04 43" src="https://github.com/fork-dev/Tracker/assets/618115/ba0a4b21-3457-42cc-9c8a-9fae2cd5904a">

**Comment by icezohu** [comment]
> @r-owen
> 
> > I think you should try SourceTree before dismissing this suggestion.
> 
> About 7 years ago I was looking for a git client for Mac. I tried ST, couldn't figure out how to work with it and in desperation I decided to make Fork.
> 
> > I find it very useful to be able to compare the current state of a file (local changes) to older commits.
> 
> Right click on a commit -> `Compare to Local Changes`
> 
> <img alt="Screenshot 2023-05-09 at 21 04 43" width="356" src="https://user-images.githubusercontent.com/618115/237195620-ba0a4b21-3457-42cc-9c8a-9fae2cd5904a.png">

wow, it's useful, didn't find this option before.

**Comment by timmb** [comment]
This is my most requested feature. I like it also because it makes it much clearer that the working tree is dirty when looking at the view. I've been using Fork a while now but I still find myself losing time by making changes without realising I had unstaged work left over from a previously interrupted task.

And as mentioned, the workflow to compare local changes to an earlier revision is the same as comparing two revisions.

**Comment by Amathlai-softs** [comment]
This is an important feature to add.  
It shows you the changes in a single snapshot view against the current branch.

Let’s not go in blind.  
You’re on branch `AAA` and you make some local changes.  

You expect to see, after `AAA`, a code diff that *can* or *will* be on top of `AAA`.  
Sometimes, you also want to compare those uncommitted changes to *another* commit or branch.

Being able to select them directly in the main screen is absolutely essential.

**Comment by danielsaidi** [comment]
I'm also coming from ST and just decided to try Fork...and unsurprisingly I desperately miss this too.
